File: dynamic_programming/linked_list.cpp
using namespace std;
#include <iostream>
#include <vector>


// Creat a Node class
class Node {
    public:
        int data;
        Node* next; // Pointer to the next node
        Node(int data) {
            this->data = data;
            this->next = NULL;
        }
};

// Create a LinkedList class
class LinkedList {
    public: 
        Node* head; // Pointer to the head of the linked list
        LinkedList() {
            this->head = NULL;
        } 

        // Add a node to the end of the linked list
        void append(int data) {
            Node* new_node = new Node(data);
            if (this->head == NULL) {
                this->head = new_node;
                return;
            }
            Node* current = this->head;
            while (current->next != NULL) {
                current = current->next;
            }
            current->next = new_node;
        }

        // Search for a node in the linked list
        Node* search(int data) {
            Node* current = this->head;
            while (current != NULL) {
                if (current->data == data) {
                    return current;
                }
                current = current->next;
            }
            return NULL;
        }

        // Insert a node after a given node
        void insert_after(Node* node, int data) {
            Node* new_node = new Node(data);
            new_node->next = node->next;
            node->next = new_node;
        }

        // Delete a node from the linked list
        void delete_node(int data) {
            Node* current = this->head;
            if (current->data == data) {
                this->head = current->next;
                delete current;
                return;
            }
            while (current->next != NULL) {
                if (current->next->data == data) {
                    Node* temp = current->next;
                    current->next = current->next->next;
                    delete temp;
                    return;
                }
                current = current->next;
            }
        }

        // Print the linked list
        void print() {
            Node* current = this->head;
            while (current != NULL) {
                cout << current->data << " ";
                current = current->next;
            }
            cout << endl;
        }

        // Reverse the linked list
        void reverse() {
            Node* prev = NULL;
            Node* current = this->head;
            Node* next = NULL;
            while (current != NULL) {
                next = current->next;
                current->next = prev;
                prev = current;
                current = next;
            }
            this->head = prev;
        }

        // Check if the linked list is a palindrome
        bool is_palindrome() {
            Node* current = this->head;
            vector<int> data;
            while (current != NULL) {
                data.push_back(current->data);
                current = current->next;
            }
            int n = data.size();
            for (int i = 0; i < n/2; i++) {
                if (data[i] != data[n-i-1]) {
                    return false;
                }
            }
            return true;
        }

        // Find the middle node of the linked list
        Node* middle_node() {
            Node* slow = this->head;
            Node* fast = this->head;
            while (fast != NULL && fast->next != NULL) {
                slow = slow->next;
                fast = fast->next->next;
            }
            return slow;
        }

        // Merge two sorted linked lists
        static LinkedList merge_sorted(LinkedList l1, LinkedList l2) {
            LinkedList merged;
            Node* current1 = l1.head;
            Node* current2 = l2.head;
            while (current1 != NULL && current2 != NULL) {
                if (current1->data < current2->data) {
                    merged.append(current1->data);
                    current1 = current1->next;
                } else {
                    merged.append(current2->data);
                    current2 = current2->next;
                }
            }
            while (current1 != NULL) {
                merged.append(current1->data);
                current1 = current1->next;
            }
            while (current2 != NULL) {
                merged.append(current2->data);
                current2 = current2->next;
            }
            return merged;
        }

        // Remove duplicates from a sorted linked list
        void remove_duplicates() {
            Node* current = this->head;
            while (current != NULL && current->next != NULL) {
                if (current->data == current->next->data) {
                    Node* temp = current->next;
                    current->next = current->next->next;
                    delete temp;
                } else {
                    current = current->next;
                }
            }
        }

        // Rotate the linked list by k positions
        void rotate(int k) {
            Node* current = this->head;
            int n = 0;
            while (current != NULL) {
                n++;
                current = current->next;
            }
            k = k % n;
            if (k == 0) {
                return;
            }
            Node* slow = this->head;
            Node* fast = this->head;
            for (int i = 0; i < k; i++) {
                fast = fast->next;
            }
            while (fast->next != NULL) {
                slow = slow->next;
                fast = fast->next;
            }
            fast->next = this->head;
            this->head = slow->next;
            slow->next = NULL;
        }
};

int main() {
    LinkedList l;
    l.append(1);
    l.append(2);
    l.append(3);
    l.append(4);
    l.append(5);
    l.append(6);
    l.append(7);
    l.append(8);
    l.print();
    l.rotate(3);
    l.print();
    l.reverse();
    l.print();
    l.insert_after(l.search(5), 9);
    l.print();
    l.delete_node(9);
    l.print();
    return 0;
}

File: dynamic_programming/linked_list.java
package dynamic_programming;

// This is a linked list implementation in Java
// The linked list is a data structure that consists of a sequence of elements, each element points to the next element in the sequence
// The linked list is a dynamic data structure, which means that the size of the list can be changed during the execution of the program
// The linked list is a linear data structure, which means that the elements are stored in a sequence and each element has a unique position in the list


// Node class
class Node {
    int data;
    Node next;

    // Constructor
    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

// LinkedList class
class LinkedList {
    Node head;

    // Constructor
    LinkedList() {
        this.head = null;
    }

    // Method to add a new node to the linked list
    public void addNode(int data) {
        Node newNode = new Node(data);

        if (head == null) {
            head = newNode;
        } else {
            Node temp = head;
            while (temp.next != null) {
                temp = temp.next;
            }
            temp.next = newNode;
        }
    }

    // Search a node in the linked list
    Node searchNode(int data) {
        Node current = head;
        while (current != null) {
            if (current.data == data) {
                return current;
            }
            current = current.next;
        }
        return null;
    }

    // Method to insert a new node after a given node
    public void insertAfter(Node prevNode, int data) {
        if (prevNode == null) {
            System.out.println("The given previous node cannot be null");
            return;
        }

        Node newNode = new Node(data);
        newNode.next = prevNode.next;
        prevNode.next = newNode;
    }

    // Method to delete a node from the linked list
    public void deleteNode(int data) {
        Node temp = head;
        Node prev = null;

        if (temp != null && temp.data == data) {
            head = temp.next;
            return;
        }

        while (temp != null && temp.data != data) {
            prev = temp;
            temp = temp.next;
        }

        if (temp == null) {
            System.out.println("The given node is not present in the linked list");
            return;
        }

        prev.next = temp.next;
    }

    // Method to reverse linked list
    public void reverse() {
        Node prevNode = null;
        Node currentNode = head;
        Node nextNode = null;
        while (currentNode != null) {
            nextNode = currentNode.next;
            currentNode.next = prevNode;
            prevNode = currentNode;
            currentNode = nextNode;
        }
        head = prevNode;
    }

    // Method to rotate the list with k times
    public void rotate(int k) {
        Node currentNode = head;
        int n = 0;
        while (currentNode != null) {
            n = n + 1;
            currentNode = currentNode.next;
        }
        k = k % n;
        if (k == 0) {
            return;
        }
        Node slow = head;
        Node fast = head;
        for (int i = 0; i < k; i++) {
            fast = fast.next;
        }
        while (fast.next != null) {
            slow = slow.next;
            fast = fast.next;
        }
        fast.next = head;
        head = slow.next;
        slow.next = null;
    }

    // Method to print the linked list
    public void printList() {
        Node temp = head;
        while (temp != null) {
            System.out.print(temp.data + " ");
            temp = temp.next;
        }
        System.out.println();
    }
}

// Main class
public class linked_list {
    public static void main(String[] args) {
        LinkedList list = new LinkedList();

        list.addNode(1);
        list.addNode(2);
        list.addNode(3);
        list.addNode(4);
        list.addNode(5);
        list.printList();
        list.insertAfter(list.searchNode(3), 9);
        list.printList();
        list.deleteNode(0);
        list.printList();
        list.deleteNode(3);
        list.printList();
        list.reverse();
        list.printList();
        list.rotate(2);
        list.printList();
    }
}